services/QuillLMS/app/queries/prompt_feedback_history.rb

```diff
@@ -18,36 +25,126 @@ def self.promptwise_sessions(activity_id)
+                pct_first_attempt_optimal: 0.0,
+                pct_first_attempt_not_optimal: 0.0
             }
 
-            prompt_hash[prompt_id][:display_name] = prompts.find(prompt_id).text
+            prompt_hash[prompt_id][:display_name] = prompts.find(prompt_session.prompt_id).text
+
+            if first_attempt_optimal?(prompt_session.attempts, prompt_session.optimals)
+              prompt_hash[prompt_id][:num_first_attempt_optimal] += 1
+            else 
+              prompt_hash[prompt_id][:num_first_attempt_not_optimal] += 1 
+            end
+
+            if consecutive_repeated_rule?(prompt_session.attempts, prompt_session.rule_uids)
+              prompt_hash[prompt_id][:num_consecutive_repeated_attempts_for_same_rule] += 1
+            end 
+
+            if non_consecutive_repeated_rule?(prompt_session.attempts, prompt_session.rule_uids)
+              prompt_hash[prompt_id][:num_non_consecutive_repeated_attempts_for_same_rule] += 1
+            end 
 
             if prompt_session.at_least_one_optimal
-              prompt_hash[prompt_id][:optimal_final_attempts] += 1
+              prompt_hash[prompt_id][:num_final_attempt_optimal] += 1
+              prompt_hash[prompt_id][:optimal_attempt_array].append prompt_session.attempt_cardinal
+            else 
+              prompt_hash[prompt_id][:num_final_attempt_not_optimal] += 1
             end
 
             prompt_hash[prompt_id][:total_responses] += prompt_session.session_count
             prompt_hash[prompt_id][:session_count] += 1
+        end
+        apply_summations(prompt_hash)
+    end
+
+    def self.apply_summations(prompt_hash)
+      prompt_hash.each do |k,v|
+        session_count = prompt_hash[k][:session_count].to_f
 
-            prompt_hash[prompt_id][:final_attempt_pct_optimal] = \
-              prompt_hash[prompt_id][:optimal_final_attempts] / prompt_hash[prompt_id][:session_count].to_f      
+        if v[:num_final_attempt_optimal] == 0
+          prompt_hash[k][:avg_attempts_to_optimal] = 0.0
+        else 
+          prompt_hash[k][:avg_attempts_to_optimal] = v[:optimal_attempt_array].sum / v[:num_final_attempt_optimal].to_f
+        end
+
+        prompt_hash[k][:pct_consecutive_repeated_attempts_for_same_rule] = \
+          v[:num_consecutive_repeated_attempts_for_same_rule] / session_count
+
+        prompt_hash[k][:pct_non_consecutive_repeated_attempts_for_same_rule] = \
+          v[:num_non_consecutive_repeated_attempts_for_same_rule] / session_count
+
+        prompt_hash[k][:pct_final_attempt_optimal] = \
+          prompt_hash[k][:num_final_attempt_optimal] / session_count      
+
+        prompt_hash[k][:pct_final_attempt_not_optimal] = \
+          prompt_hash[k][:num_final_attempt_not_optimal] / session_count
+          
+        prompt_hash[k][:pct_first_attempt_optimal] = v[:num_first_attempt_optimal] / session_count
+        prompt_hash[k][:pct_first_attempt_not_optimal] = v[:num_first_attempt_not_optimal] / session_count
```

Comment:
Yeah, I agree, especially because these "percentages" are actually decimals right now and have to get multiplied by *100 for display on the frontend anyway.

Reply:
sure